core/include/moveit/task_constructor/properties/serialize.hpp

```diff
@@ -0,0 +1,184 @@
+template <typename T, typename = std::ostream&>
+struct hasInsertionOperator : std::false_type
+{};
+
+template <typename T>
+struct hasInsertionOperator<T, decltype(std::declval<std::ostream&>() << std::declval<T>())> : std::true_type
+{};
+
+template <typename T>
+constexpr bool hasInsertionOperator_v = hasInsertionOperator<T>::value;
+
+template <typename T, typename = std::istream&>
+struct hasExtractionOperator : std::false_type
+{};
+
+template <typename T>
+struct hasExtractionOperator<T, decltype(std::declval<std::istream&>() >> std::declval<T&>())> : std::true_type
+{};
+
+template <typename T>
+constexpr bool hasExtractionOperator_v = hasExtractionOperator<T>::value;
+
+class PropertySerializerBase
+{
+public:
+	using SerializeFunction = std::string (*)(const boost::any&);
+	using DeserializeFunction = boost::any (*)(const std::string&);
+
+	static std::string dummySerialize(const boost::any& /*unused*/) { return ""; }
+	static boost::any dummyDeserialize(const std::string& /*unused*/) { return boost::any(); }
+
+protected:
+	static bool insert(const std::type_index& type_index, const std::string& type_name, SerializeFunction serialize,
+	                   DeserializeFunction deserialize);
+};
+
+template <typename T>
+class PropertySerializer : public PropertySerializerBase
+{
+protected:
+	/// a set of mutually exclusive conditions
+	template <typename U>
+	static constexpr bool IsROSMessage{ ros::message_traits::IsMessage<U>::value };
+
+	template <typename U>
+	static constexpr bool IsStreamSerializable{ !IsROSMessage<U> && hasInsertionOperator_v<U> };
+
+	template <typename U>
+	static constexpr bool IsStreamDeserializable{ !IsROSMessage<U> && hasExtractionOperator_v<U> };
+
+	template <typename U>
+	static constexpr bool IsString{ std::is_same<U, std::string>::value };
+
+	template <typename U>
+	static constexpr bool IsNotDeserializable{ !IsROSMessage<U> && !IsStreamSerializable<U> };
+
+public:
+	PropertySerializer() { this->insert(typeid(T), this->typeName<T>(), &serialize, &deserialize); }
+
+	template <typename Q = T>
+	static std::enable_if_t<!IsROSMessage<Q>, std::string> typeName() {
+		return typeid(T).name();
+	}
+
+	template <typename Q = T>
+	static std::enable_if_t<IsROSMessage<Q>, std::string> typeName() {
+		return ros::message_traits::DataType<T>::value();
+	}
+
+	/** ROS messages are serialized using the provided binary encoding */
+	template <typename Q = T>
+	static std::enable_if_t<IsROSMessage<Q>, std::string> serialize(const boost::any& value) {
+		Q message{ boost::any_cast<Q>(value) };
+		uint32_t serial_size = ros::serialization::serializationLength(message);
+		std::unique_ptr<uint8_t[]> buffer(new uint8_t[serial_size]);
+		ros::serialization::OStream stream(buffer.get(), serial_size);
+		ros::serialization::serialize(stream, message);
+		return reinterpret_cast<const char*>(base64::encode(buffer.get(), serial_size).c_str());
```

Comment:
Why do you think we need Base64 encoding here? A `std::string` can contain any `char` - in contrast to `char*`, which cannot include `\0` as this already encodes the end of the string. However, a std::string's length is encoded separately.

Reply:
My new commits prove that this isn't required.